File: solutions/elixir/154-all-nodes-distance-k-in-binary-tree.ex
# Problem 154: All Nodes Distance K in Binary Tree (LeetCode 863)
# Difficulty: Med
# Language: Elixir
# 
defmodule DistanceK do
  def solve(root, target, k) do
    parent_map = build_parents(root, nil, %{})
    bfs([target], 0, k, MapSet.new([target.val]), parent_map)
  end
  defp build_parents(nil, _, map), do: map
  defp build_parents(%{val: v, left: l, right: r}, parent, map) do
    map = Map.put(map, v, parent)
    map = build_parents(l, v, map)
    build_parents(r, v, map)
  end
  defp bfs(queue, dist, k, _, _) when dist == k, do: Enum.map(queue, & &1.val)
  defp bfs([], _, _, _, _), do: []
  defp bfs(queue, dist, k, visited, pm) do
    {next, visited} = Enum.reduce(queue, {[], visited}, fn node, {nxt, vis} ->
      neighbors = [node.left, node.right, pm[node.val]]
        |> Enum.reject(&(is_nil(&1) or MapSet.member?(vis, &1)))
      {nxt ++ neighbors, Enum.reduce(neighbors, vis, &MapSet.put(&2, &1))}
    end)
    bfs(next, dist + 1, k, visited, pm)
  end
end

File: solutions/typescript/154-all-nodes-distance-k-in-binary-tree.ts
/**
 * Problem 154: All Nodes Distance K in Binary Tree (LeetCode 863)
 * Difficulty: Med
 * Language: TypeScript
 */
interface TreeNode { val: number; left: TreeNode | null; right: TreeNode | null; }
function distanceK(root: TreeNode | null, target: TreeNode, k: number): number[] {
  const parentMap = new Map<TreeNode, TreeNode | null>();
  const buildParent = (node: TreeNode | null, parent: TreeNode | null): void => {
    if (!node) return;
    parentMap.set(node, parent);
    buildParent(node.left, node);
    buildParent(node.right, node);
  };
  buildParent(root, null);
  const bfs = (start: TreeNode, k: number): number[] => {
    const visited = new Set<TreeNode>();
    let queue: TreeNode[] = [start];
    visited.add(start);
    let dist = 0;
    while (queue.length > 0 && dist < k) {
      queue = queue.reduce<TreeNode[]>((next, node) => {
        const neighbors = [node.left, node.right, parentMap.get(node) ?? null]
          .filter((n): n is TreeNode => n !== null && !visited.has(n));
        neighbors.forEach(n => visited.add(n));
        return [...next, ...neighbors];
      }, []);
      dist++;
    }
    return queue.map(n => n.val);
  };
  return bfs(target, k);
}
